### PR TITLE
Swizzle support for Vectors and float

### PR DESCRIPTION
**Pull request by amerkoleci**
Hi,
As promised I've added swizzle support with t4 automatic template generation.
I've added also some code into TestProject to show the logic.

### CODE REVIEW

File: src/ShaderGen.Primitives/ShaderSwizzle.tt [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+<#@ import namespace="System" #>
+<#@ import namespace="System.Linq" #>
+<#@ import namespace="System.Collections.Generic"#>
+<#@ import namespace="System.Runtime.InteropServices" #>
+<#@ import namespace="System.Diagnostics" #>
+using System.Numerics;
+
+namespace ShaderGen
+{
+	public static class ShaderSwizzle
```
Comment (mellinoe) [inline src/ShaderGen.Primitives/ShaderSwizzle.tt]: Can you replace all the tabs with spaces in this file?

File: src/ShaderGen/Glsl450Backend.cs [hunk 2]
```diff
@@ -91,5 +91,6 @@ protected override void EmitGlPositionCorrection(StringBuilder sb)
         {
             sb.AppendLine($"        gl_Position.y = -gl_Position.y; // Correct for Vulkan clip coordinates");
         }
+
```
Comment (mellinoe) [inline src/ShaderGen/Glsl450Backend.cs]: Undo

**Comment by mellinoe** [comment]
Hey @amerkoleci , thanks for the help!

Instead of plumbing this through the various backends, I would prefer to have the function translation done the same way that the other ShaderBuiltin functions are translated right now. I refactored the "KnownFunctions" classes a bit to allow this. You should be able to define a new `TypeInvocationTranslator` which handles swizzle calls, and just looks at the name of the method to figure out which element accesses to emit.

Aside from that, I'd prefer if the C# swizzle methods used upper-case element letters ("XXYY()", "WZYX", etc.) rather than lower-case, since the C# fields are also upper-case. I'd also prefer to just have the XYZW swizzles at first and not the RGBA swizzles -- I think they could be confusing in C# if you had a different component order for your pixel data.

Let me know if you need help getting this working, I'll probably not be working on this much until next week after tonight.

**Comment by amerkoleci** [comment]
Hey @mellinoe ,
I will fix the notes you reported, I've added the RGBA swizzle as they are nativelly supported by the shader language, but if you prefer I can remove them, I will convert to upper case the swizzle logic.

**Comment by amerkoleci** [comment]
Should be fine now. About TypeInvocationTranslator I won't be able to do so, maybe you can merge and refactor later :)

**Comment by mellinoe** [comment]
Thanks! I will most likely pull your branch down locally and then refactor the backends a bit before pushing this up to `master`. Hopefully I can get to that tonight.

**Comment by mellinoe** [comment]
@amerkoleci I pulled your changes down locally and made an extra integration commit on top. It should be in master now.

**Pull request closed without merging**